data/dataset.py

```diff
@@ -0,0 +1,126 @@
+import os.path
+import itertools
+
+import torch
+import fortepyan as ff
+from tqdm import tqdm
+from tokenizer import Tokenizer
+from datasets import load_dataset
+from prepare_dataset import process_record
+
+from data.quantizer import MidiQuantizer
+
+
+class TokenizedMidiDataset:
+    def __init__(
+        self,
+        split: str = "train",
+        n_dstart_bins: int = 3,
+        n_duration_bins: int = 10,
+        n_velocity_bins: int = 10,
+        sequence_len: int = 128,
+        device: str = "cpu",
+    ):
+        self.sequence_len = sequence_len
+        self.device = device
+
+        self.quantizer = MidiQuantizer(
+            n_dstart_bins=n_dstart_bins,
+            n_duration_bins=n_duration_bins,
+            n_velocity_bins=n_velocity_bins,
+        )
+
+        self.tokenizer_src = Tokenizer(keys=["pitch", "dstart_bin"])
+        self.tokenizer_tgt = Tokenizer(keys=["duration_bin", "velocity_bin"])
```

Comment:
My initial idea was about predicting a high resolution velocity information, from low resolution quantization:

```py
        # 3 * 3 * 3 * 88 = 2376 tokens
        self.quantizer = MidiQuantizer(
            n_dstart_bins=3,
            n_duration_bins=3,
            n_velocity_bins=3,
        )
        # Source tokenizer should use all (low resolution) information
        self.tokenizer_src = Tokenizer(keys=["pitch", "dstart_bin", "duration_bin", "velocity_bin"])

        # This can have up to 128 tokens, which would be one token per every possible velocity value
        self.tokenizer_tgt = VelocityTokenizer()
```

So I think it might be better to have dedicated tokenizers for ins and outs of the problem 🤔

Reply:
Do I understand it correctly - we want to have all the quantized information on input and predict only high resolution velocity?
My initial thought was to enable the model to predict higher resolution velocity and duration based on low resolution dstart and pitch. I think it would be interestiong to see waht predictions would the transformer come up with. But we can do velocities first, it will probably work a lot better.

Reply:
+ I can still use the same class, I just need to set keys as ['velocity'] and pass original notes df

Reply:
> Do I understand it correctly - we want to have all the quantized information on input and predict only high resolution velocity?

That's right

> velocities first, it will probably work a lot better

Yes, the idea is to start with an "easy" problem and if that works look for a way to get it to work on more complex problems :)

For the tokenizers, I think having separate classes will be more explicit about what's going on - looks like we'll have to test multiple different approaches to tokenization, and I don't want to build a single class doing everything for this 🤔

Reply:
Okay, I’ll work on it